From cb55884f23d95b4d5806d070e6219e9bc56c8db4 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Thu, 06 Apr 2023 15:12:17 +0300
Subject: [PATCH] [SW-140559] gaudi2: boot_info: copy from perst mem

This patch adds support to copy boot_info from
persistent boot_info memory to current boot_info memory
in ppboot.

Earlier this step was done in early preboot main().
But as boot_info is required even before preboot main
by driver(s), so this is added in ppboot.

Change-Id: I357dacda0792cf70df8660c96d342733832b793b
---

diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index cdef38d..406f4ab 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -311,6 +311,7 @@
 __weak void ppboot_plls_init(void) {}
 __weak void hl_pcie_enable_elbi_access(void) {}
 __weak void ppboot_tpm_extend(void) {}
+__weak void ppboot_boot_info_init(void) {}
 
 void ppboot_main(void)
 {
@@ -360,6 +361,7 @@
 
 	preboot_delay();
 
+	ppboot_boot_info_init();
 	/* SRAM PLLs must be configured before running BIST operations */
 	hl_sram_self_repair();
 
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index cf529b6..e2c7f9e 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -1051,3 +1051,12 @@
 	if (boot_info_sync_data())
 		preboot_puts("Failed to merge old boot info\n");
 }
+
+void ppboot_boot_info_init(void)
+{
+	u32 rst_src = hl_rst_src_get();
+
+	/* If it is not COLD_RST, copy boot info from persistent memory to SPSRAM */
+	if (rst_src != HL_COLD_RST)
+		boot_info_copy_from_perst_mem();
+}
diff --git a/include/preboot.h b/include/preboot.h
index 2aa2603..ef75fcd 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -268,6 +268,7 @@
 void preboot_security_sp_sram_init(void);
 u32 hl_get_pll_div_fix(void);
 void preboot_boot_info_init(void);
+void ppboot_boot_info_init(void);
 
 #endif /* __ASSEMBLY__ */
 
